ext/dom/document.c

```diff
@@ -1557,15 +1729,26 @@ PHP_METHOD(DOMDocument, saveXML)
 			php_error_docref(NULL, E_WARNING, "Could not create save context");
 			RETURN_FALSE;
 		}
-		if (UNEXPECTED(xmlSaveDoc(ctxt, docp) < 0)) {
-			(void) xmlSaveClose(ctxt);
-			xmlBufferFree(buf);
-			php_error_docref(NULL, E_WARNING, "Could not save document");
-			RETURN_FALSE;
+		if (php_dom_follow_spec_intern(intern)) {
+			xmlCharEncodingHandlerPtr handler = xmlFindCharEncodingHandler((const char *) docp->encoding);
+			xmlOutputBufferPtr out = xmlOutputBufferCreateBuffer(buf, handler);
+			if (EXPECTED(out != NULL)) {
+				status = dom_xml_serialize(ctxt, out, (xmlNodePtr) docp, format);
+				status |= xmlOutputBufferFlush(out);
+				status |= xmlOutputBufferClose(out);
+			} else {
+				xmlCharEncCloseFunc(handler);
+			}
+		} else {
+			status = xmlSaveDoc(ctxt, docp);
 		}
-		(void) xmlSaveFlush(ctxt);
 		(void) xmlSaveClose(ctxt);
 	}
+	if (UNEXPECTED(status < 0)) {
+		xmlBufferFree(buf);
+		php_error_docref(NULL, E_WARNING, "Could not save document");
+		RETURN_FALSE;
+	}
```

Comment:
Should this possibly throw?

Reply:
That's true, could use one of the SPL exceptions but I'd rather not, will have a think about this